python/ml4ir/base/model/layers/sentence_transformers.py

```diff
@@ -0,0 +1,158 @@
+from tensorflow.keras.layers import Layer, Dense
+from transformers import TFAutoModel, TFBertTokenizer
+from sentence_transformers import SentenceTransformer
+
+# NOTE: We set device CPU for the torch backend so that the sentence-transformers model does not use GPU resources
+torch.device("cpu")
+
+
+class SentenceTransformerLayerKey:
+    """Stores the names of the sentence-transformer model layers"""
+    TRANSFORMER = "sentence_transformers.models.Transformer"
+    POOLING = "sentence_transformers.models.Pooling"
+    DENSE = "sentence_transformers.models.Dense"
+    NORMALIZE = "sentence_transformers.models.Normalize"
+
+
+class SentenceTransformerWithTokenizerLayer(Layer):
+    """
+    Converts a string tensor into embeddings using sentence transformers
+    by first tokenizing the string tensor and then passing through the transformer model
+
+    Some of this code is inspired from -> https://www.philschmid.de/tensorflow-sentence-transformers
+    """
+
+    def __init__(self,
+                 name="sentence_transformer",
+                 model_name_or_path: str = "intfloat/e5-base",
+                 trainable: bool = False,
+                 **kwargs):
+        """
+        Parameters
+        ----------
+        name: str
+            Layer name
+        model_name_or_path: str
+            Name or path to the sentence transformer embedding model
+        trainable: bool
+            Finetune the pretrained embedding model
+        kwargs:
+            Additional key-value args that will be used for configuring the layer
+        """
+        super().__init__(name=name, **kwargs)
+
+        self.model_name_or_path = Path(model_name_or_path)
+        if not Path(model_name_or_path).exists():  # The user provided a model name
+            # If the sentence_transformer model files are not present, we initialize it to trigger a download
+            st_model = SentenceTransformer(model_name_or_path)
+
+            self.model_name_or_path = Path(
+                torch.hub._get_torch_home()) / "sentence_transformers" / model_name_or_path.replace("/", "_")
+            if not self.model_name_or_path.exists():
+                raise FileNotFoundError(
+                    f"{self.model_name_or_path} does not exist. Verify the `model_name_or_path` argument")
+
+            del st_model
+
+        # Load the modules.json config to add custom model layers
+        self.modules = json.load(open(self.model_name_or_path / "modules.json"))
+        self.module_names = [module["name"] for module in self.modules]
+
+        self.trainable = trainable
+
+        # Define tokenizer as part of the tensorflow graph
+        self.tokenizer = TFBertTokenizer.from_pretrained(self.model_name_or_path, **kwargs)
+
+        self.transformer_model = None
+        self.dense = None
+        self.apply_dense = False
+        self.normalize_embeddings = False
+        self.pool_embeddings = False
+        for module in self.modules:
```

Comment:
Is that's the fix to the issue of loading different models? @lastmansleeping I'm curious what was the issue